heimlig/src/hsm/workers/ecc_worker.rs

```diff
@@ -0,0 +1,178 @@
+        let locked_key_store = self.key_store.lock().await;
+        let key_info = locked_key_store.deref().get_key_info(key_id);
+        match key_info {
+            Err(e) => Response::Error {
+                client_id,
+                request_id,
+                error: Error::KeyStore(e),
+            },
+            Ok(key_info) => {
+                let key_exists = locked_key_store.deref().is_stored(key_id);
+                if key_exists && !key_info.permissions.overwrite {
+                    return Response::Error {
+                        client_id,
+                        request_id,
+                        error: Error::KeyStore(keystore::Error::NotAllowed),
+                    };
+                }
+                if !key_info.ty.is_asymmetric() {
+                    return Response::Error {
+                        client_id,
+                        request_id,
+                        error: Error::KeyStore(keystore::Error::InvalidKeyType),
+                    };
+                }
+                match key_info.ty {
+                    KeyType::EccKeypairNistP256 => {
+                        self.generate_key_pair_internal::<NistP256>(
+                            locked_key_store,
+                            client_id,
+                            request_id,
+                            key_info,
+                        )
+                        .await
+                    }
+                    KeyType::EccKeypairNistP384 => {
+                        self.generate_key_pair_internal::<NistP384>(
+                            locked_key_store,
+                            client_id,
+                            request_id,
+                            key_info,
+                        )
+                        .await
+                    }
+                    _ => Response::Error {
+                        client_id,
+                        request_id,
+                        error: Error::KeyStore(keystore::Error::InvalidKeyType),
+                    },
+                }
+            }
+        }
+    }
+
+    async fn generate_key_pair_internal<C>(
+        &mut self,
+        mut locked_key_store: MutexGuard<'_, M, &mut (dyn KeyStore + Send)>,
+        client_id: ClientId,
+        request_id: RequestId,
+        key_info: KeyInfo,
+    ) -> Response<'data>
+    where
+        C: CurveArithmetic,
+        AffinePoint<C>: FromEncodedPoint<C> + ToEncodedPoint<C>,
+        FieldBytesSize<C>: ModulusSize,
+    {
+        let (public_key, private_key) =
+            generate_key_pair::<_, C>(self.rng.lock().await.deref_mut());
+
+        // Write public key
+        let mut public_key_bytes = [0u8; KeyType::MAX_PUBLIC_KEY_SIZE];
+        let public_key_bytes = &mut public_key_bytes[..key_info.ty.public_key_size()];
+        let public_key_encoded = &public_key.to_encoded_point(false);
+        let public_key_coordinates = public_key_encoded.coordinates();
+        match public_key_coordinates {
```

Comment:
Sorry, for not proposing it in the first review: 
It might be easier to use `EncodedPoint` to retrieve the public key bytes. 
```
let point: EncodedPoint<C> = public_key.into();
let point_bytes = point.as_bytes();
// point bytes contains 1 byte tag (uncompressed = 4) and concatenated x and y coordinates
```

To restore the public key.
```
let p = EncodedPoint::<C>::from_untagged_bytes(bytes.into());
let pk_res = PublicKey::from_encoded_point(&p);
```

Reply:
Calling `let point: EncodedPoint<C> = public_key.into();` gives you indeed an sec1 encoded point but it is not clear to me if that point will always be uncompressed. It seems to rely on `C::COMPRESS_POINTS` which can be different for every curve. I however managed to simplify the code by using `as_bytes()` as you suggested.

Restoration of keys will be helpful when implementing signing and verification ;)